lib/ASRouter.jsm

```diff
@@ -343,6 +347,57 @@ class _ASRouter {
     }
   }
 
+  async addImpression(message) {
+    // Don't store impressions for messages that don't include any limits on frequency
+    if (!message.caps) {
+      return;
+    }
+    await this.setState(state => {
+      const impressions = {...state.impressions};
+      const impressionsForMessage = impressions[message.id] ? [...impressions[message.id]] : [];
+      impressions[message.id] = impressionsForMessage;
+      impressionsForMessage.push(Date.now());
+      this._storage.set("impressions", impressions);
+      return {impressions};
+    });
+  }
+
+  getLongestPeriod(message) {
+    if (!message.caps || !message.caps.custom) {
+      return null;
+    }
+    return message.caps.custom.sort((a, b) => b.period - a.period)[0].period;
+  }
+
+  async cleanupImpressions() {
```

Comment:
This is the cleanup code here @ncloudioj

Reply:
Sure, will take a look.